### ISSUE TITLE
 reportIncompatibleVariableOverride  even though the override type is a subclass of the base type.

### ISSUE THREAD

**Issue by amarion35** [microsoft/pyright#7840]
## Environment data

-   Language Server version: 2024.4.1
-   OS and version: WSL2 - Ubuntu 20.04.6
-   Python version (& distribution if applicable, e.g. Anaconda): 3.11

## Code Snippet

```python
class ASettings:
    a: int = 1


class BSettings(ASettings):
    b: int = 2


class AService:
    settings: ASettings

    def __init__(self, settings: ASettings) -> None:
        self.settings = settings

    def print_a(self) -> None:
        print(self.settings.a)


class BService(AService):
    settings: BSettings #"settings" overrides symbol of same name in class "AService". Variable is mutable so its type is invariant. Override type "BSettings" is not the same as base type "ASettings". Pylance[reportIncompatibleVariableOverride](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportIncompatibleVariableOverride)

    def __init__(self, settings: BSettings) -> None:
        super().__init__(settings)

    def print_b(self) -> None:
        print(self.settings.b)
```

## Expected behavior

Indeed type BSettings is not the same as type ASettings. But according to [this answer](https://github.com/microsoft/pylance-release/issues/2303#issuecomment-1024772968) "Subclasses can override it with a narrow type if they'd like" so, since BSettings is a subtype of ASettings, this type override should not be considered as incompatible.

## Actual behavior

Pylance gives a reportIncompatibleVariableOverride error.

**Comment by erictraut** [comment]
Pyright's behavior here is correct. What you're doing here is not type safe. It is type safe to override a read-only attribute in a covariant manner, but you're attempting to override a writable (mutable) attribute in a covariant manner. In your example, `AService` (which is the base class of `BService`) can set `settings` to a value that is an instance of `ASettings`, but this would make `settings` incompatible with a `BService` instance.

The answer you quoted is outdated. The Python typing community has since concluded that Python type checkers should be type safe in this regard.

There has been some discussion about extending the recently-added `typing.ReadOnly` special form (which is currently used only for `TypedDict` definitions) to apply more broadly to non-TypedDict classes. This would allow one to specify a read-only attribute which would make it type safe to override in a covariant manner. This hasn't yet made it to the PEP stage though.

In the meantime, if you want to override a mutable class variable in a covariant manner, you will need to either use a `# type: ignore` or disable pyright's `reportIncompatibleVariableOverride` check. You can do this globally for your project or on a file-by-file basis by adding a `# pyright: reportIncompatibleVariableOverride=false` comment at the top of the file.

Closing this issue because pyright is working correctly, and this isn't a bug.
